keps/0021-kudo-upgrade.md

```diff
@@ -0,0 +1,271 @@
+
+### Goals
+
+- How updates of KUDO are executed
+  - Updates of CRDs
+  - Updates of the KUDO manager
+  - Updates of prerequisites (Namespaces, RoleBindings, ServiceAccounts, etc.)
+- Interoperability
+  - How and if multiple versions of CRDs are maintained
+  - How a version of KUDO CLI work with older and newer CRD versions
+- How to handle operators that are not supported by a new KUDO version
+
+### Non-Goals
+
+- Updates/Upgrades of operators and operator versions itself
+- Versioning of KUDO itself
+- Multi-Tenancy, i.e. running multiple KUDO installations on the same cluster
+
+
+### Current State
+
+At the moment, KUDO does not provide any migration capabilities and needs a clean installation to use a new version.
+
+## Open Questions
+- Lowest supported K8s version
+- Do we want to support downgrades? 
+- Split versions between KUDO manager and KUDO CLI?
+
+## Proposal
+
+### KUDO Prerequisites
+Expected update frequency: Low  
+Versioned: No, but closely tied to KUDO manager
+
+The KUDO manager has a set of prerequisites that need to be present for the manager to run successfully. They are
+the least likely to change, but probably the most specific. If we make changes here, we need to implement custom
+migration code.
+
+- Current prereqs
+  - Namespace
+  - Service Account
+  - Role Bindings
+  - Secret
+
+#### Proposal for update process 
+Integrated into `kudo init --upgrade`
+
+Write specific migration code that targets a KUDO manager version range and executes manual migration steps. 
+
+Each migration should have a validate-step that checks if the migration is possible.
+
+**Alternative update process**
+- Can we just delete them all and reinstall them? Probably not
+
+### KUDO Manager
+Expected update frequency: High  
+Versioned: Yes
+
+The KUDO Manager is defined by an image version in a deployment set. To update, the deployment must be updated. The 
+manager is closely tied to the CRDs, but not to the CLI. When CRDs are updated, the Manager will most likely also
+need to be updated. 
+
+#### Proposal for update process
+Integrated into `kudo init --upgrade`
+
+- Use semantic versioning for the manager binary
+  - ToBeDiscussed: Use different versioning for Manager and CLI?
+- The manager needs to be shut down before updating the CRDs
+  - As soon as the new CRD is installed and marked as "stored"
+  - Without WebHook Conversion, even if the new CRD version is "served", the old content is returned without any conversion
+- Question: Do we need to ensure that the manager is not doing meaningful work at the moment, or can we just update the deployment?
+
+### CRDs
+Expected update frequency: Medium  
+Versioned: Yes, with a CRD-Version
+
+The CRDs are used to store installed operators and running instances. New features will regularly require us to add new
+fields. 
+- Existing CRs need to be migrated to new versions
+- K8s CRD support
+  - Which K8s version do we want/need to support?
```

Comment:
I guess this is kind of given by what kubernetes version we support in which version of manager? Maybe we should somehow define that as well?

Reply:
Well, it mostly comes down to: Do we want to use conversion webhooks or not.